backend/src/controller/chatController.ts

```diff
@@ -187,38 +192,79 @@ async function handleChatToGPT(req: OpenAiChatRequest, res: Response) {
 				chatModel
 			);
 		}
+	} catch (error) {
+		const errorMessage =
+			error instanceof Error ? error.message : 'Failed to get chatGPT reply';
+		handleErrorGettingReply(req, res, currentLevel, chatResponse, errorMessage);
+		return;
+	}
 
-		// if the reply was blocked then add it to the chat history
-		if (chatResponse.defenceReport.isBlocked) {
-			req.session.levelState[currentLevel].chatHistory.push({
-				completion: null,
-				chatMessageType: CHAT_MESSAGE_TYPE.BOT_BLOCKED,
-				infoMessage: chatResponse.defenceReport.blockedReason,
-			});
-		} else if (!chatResponse.reply || chatResponse.reply === '') {
-			throw new Error('Failed to get chatGPT reply');
-		}
+	// more error handling
+	if (chatResponse.openAIErrorMessage) {
+		handleErrorGettingReply(
+			req,
+			res,
+			currentLevel,
+			chatResponse,
+			simplifyOpenAIErrorMessage(chatResponse.openAIErrorMessage)
+		);
+		return;
+	} else if (!chatResponse.reply || chatResponse.reply === '') {
+		handleErrorGettingReply(
+			req,
+			res,
+			currentLevel,
+			chatResponse,
+			'Failed to get chatGPT reply'
+		);
+		return;
+	}
 
-		// update sent emails
-		chatResponse.sentEmails =
-			req.session.levelState[currentLevel].sentEmails.slice(numSentEmails);
-	} catch (error) {
-		// add error message to chat history
+	if (chatResponse.defenceReport.isBlocked) {
 		req.session.levelState[currentLevel].chatHistory.push({
 			completion: null,
-			chatMessageType: CHAT_MESSAGE_TYPE.ERROR_MSG,
-			infoMessage: 'Failed to get chatGPT reply',
+			chatMessageType: CHAT_MESSAGE_TYPE.BOT_BLOCKED,
+			infoMessage: chatResponse.defenceReport.blockedReason,
 		});
-		console.error(error);
-
-		handleChatError(res, chatResponse, true, 'Failed to get chatGPT reply');
-		return;
 	}
-	// log and send the reply with defence report
+
+	// update sent emails
+	chatResponse.sentEmails =
+		req.session.levelState[currentLevel].sentEmails.slice(numSentEmails);
+
 	console.log(chatResponse);
 	res.send(chatResponse);
 }
 
+function simplifyOpenAIErrorMessage(openAIErrorMessage: string) {
+	if (openAIErrorMessage.startsWith('429')) {
+		const tryAgainMessage = openAIErrorMessage
+			.split('. ')
+			.find((sentence) => sentence.includes('Please try again in'));
```

Comment:
oh, um. I actually didn't check this. That was my assumption. Let me investigate...

Reply:
Ok. I know that the timeout changes. [Here's an example](https://community.openai.com/t/rate-limit-reached-for-10ktpm-200rpm/381878) of a different timeout. I suppose it's different per membership. But also, because there is a rate limit per minute distinct from the rate limit per day, I expect that the timeouts will reflect these. i.e. if you've exceeded the limit per day it would say (I hope) "please try again tomorrow"

Reply:
Ah cool, thanks for confirming!